applications/vanilla/models/class.discussionmodel.php

```diff
@@ -1167,6 +1159,67 @@ public function calculate(&$discussion) {
         $this->fireEvent('SetCalculatedFields');
     }
 
+    /**
+     * Calculate if the user has read all the Comments in a discussion. The data in the Discussion table
+     * is more likely to be reliable because when Comments are deleted we update the Discussion table
+     * but not the UserDiscussion table.
+     *
+     * @param int $discussionCommentCount Number of Comments according to the Discussion table.
+     * @param mixed $discussionLastCommentDate Date of last Comment according to the Discussion table.
+     * @param int $userReadComments Number of Comments the user has read according to the UserDiscussion table.
+     * @param mixed $userLastReadDate Date of last Comment read according to the UserDiscussion table.
+     * @return array
+     */
+    public static function determineReadStatusAndUnreadCount(
+        int $discussionCommentCount,
+        $discussionLastCommentDate,
+        int $userReadComments,
+        $userLastReadDate
+    ) {
+        $isRead = true;
+        $unreadCommentCount = $discussionCommentCount - $userReadComments;
+        if ($discussionLastCommentDate > $userLastReadDate) {
+            $isRead = false;
+            if ($discussionCommentCount > 0 && $userReadComments >= $discussionCommentCount) {
+                $unreadCommentCount = 1;
+            }
+        }
+
+        if ($userLastReadDate > $discussionLastCommentDate && $unreadCommentCount > 0) {
+            $unreadCommentCount = 0;
+        }
+
+        if (!$userReadComments) {
+            $unreadCommentCount = $discussionCommentCount;
+        }
+
+        if ($unreadCommentCount < 0 && !$isRead) {
+            $unreadCommentCount = 1;
+        }
+
+        if ($unreadCommentCount < 0 && $isRead) {
+            $unreadCommentCount = 0;
+        }
+
+        return [$isRead, $unreadCommentCount];
+    }
+
+    /**
+     * Decide which date we want to consider as the date last viewed, either when the
+     * user has marked the Category read or when the user last viewed the Discussion.
+     *
+     * @param mixed $discussionLastViewedDate Date discussion was last viewed.
+     * @param mixed $categoryLastViewedDate Date category was marked viewed.
+     * @return mixed Returns most recent date.
+     */
+    public static function determineLastViewedDate($discussionLastViewedDate, $categoryLastViewedDate) {
+        if ($discussionLastViewedDate === null || $discussionLastViewedDate < $categoryLastViewedDate) {
```

Comment:
`empty($discussionLastViewedDate)` is more robust. Also, you want to account for either date being empty.